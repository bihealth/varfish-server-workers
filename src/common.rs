//! Common functionality.

use std::{
    fs::File,
    io::{BufRead, BufReader, BufWriter, Write},
    ops::Range,
    path::Path,
};

use byte_unit::Byte;
use clap_verbosity_flag::{InfoLevel, Verbosity};

use clap::Parser;
use flate2::{bufread::MultiGzDecoder, write::GzEncoder, Compression};
use hgvs::static_data::Assembly;
use indexmap::IndexMap;

/// Commonly used command line arguments.
#[derive(Parser, Debug)]
pub struct Args {
    /// Verbosity of the program
    #[clap(flatten)]
    pub verbose: Verbosity<InfoLevel>,
}

/// Helper to print the current memory resident set size via `tracing`.
pub fn trace_rss_now() {
    let me = procfs::process::Process::myself().unwrap();
    let page_size = procfs::page_size();
    tracing::debug!(
        "RSS now: {}",
        Byte::from_bytes((me.stat().unwrap().rss * page_size) as u128).get_appropriate_unit(true)
    );
}

/// Definition of canonical chromosome names.
pub const CHROMS: &[&str] = &[
    "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17",
    "18", "19", "20", "21", "22", "X", "Y", "M",
];

/// Build mapping of chromosome names to chromosome counts.
pub fn build_chrom_map() -> IndexMap<String, usize> {
    let mut result = IndexMap::new();
    for (i, &chrom_name) in CHROMS.iter().enumerate() {
        result.insert(chrom_name.to_owned(), i);
        result.insert(format!("chr{chrom_name}").to_owned(), i);
    }
    result.insert("x".to_owned(), 22);
    result.insert("y".to_owned(), 23);
    result.insert("chrx".to_owned(), 22);
    result.insert("chry".to_owned(), 23);
    result.insert("mt".to_owned(), 24);
    result.insert("m".to_owned(), 24);
    result.insert("chrmt".to_owned(), 24);
    result.insert("chrm".to_owned(), 24);
    result.insert("MT".to_owned(), 24);
    result.insert("chrMT".to_owned(), 24);
    result
}

/// Transparently open a file with gzip decoder.
pub fn open_read_maybe_gz<P>(path: P) -> Result<Box<dyn BufRead>, anyhow::Error>
where
    P: AsRef<Path>,
{
    if path.as_ref().extension().map(|s| s.to_str()) == Some(Some("gz")) {
        tracing::trace!("Opening {:?} as gzip for reading", path.as_ref());
        let file = File::open(path)?;
        let bufreader = BufReader::new(file);
        let decoder = MultiGzDecoder::new(bufreader);
        Ok(Box::new(BufReader::new(decoder)))
    } else {
        tracing::trace!("Opening {:?} as plain text for reading", path.as_ref());
        let file = File::open(path).map(BufReader::new)?;
        Ok(Box::new(BufReader::new(file)))
    }
}

/// Transparently opena  file with gzip encoder.
pub fn open_write_maybe_gz<P>(path: P) -> Result<Box<dyn Write>, anyhow::Error>
where
    P: AsRef<Path>,
{
    if path.as_ref().extension().map(|s| s.to_str()) == Some(Some("gz")) {
        tracing::trace!("Opening {:?} as gzip for writing", path.as_ref());
        let file = File::create(path)?;
        let bufwriter = BufWriter::new(file);
        let encoder = GzEncoder::new(bufwriter, Compression::default());
        Ok(Box::new(encoder))
    } else {
        tracing::trace!("Opening {:?} as plain text for writing", path.as_ref());
        let file = File::create(path)?;
        Ok(Box::new(file))
    }
}

// Compute reciprocal overlap between two ranges.
pub fn reciprocal_overlap(lhs: Range<i32>, rhs: Range<i32>) -> f32 {
    let lhs_b = lhs.start;
    let lhs_e = lhs.end;
    let rhs_b = rhs.start;
    let rhs_e = rhs.end;
    let ovl_b = std::cmp::max(lhs_b, rhs_b);
    let ovl_e = std::cmp::min(lhs_e, rhs_e);
    if ovl_b >= ovl_e {
        0f32
    } else {
        let ovl_len = (ovl_e - ovl_b) as f32;
        let x1 = ovl_len / (lhs_e - lhs_b) as f32;
        let x2 = ovl_len / (rhs_e - rhs_b) as f32;
        x1.min(x2)
    }
}

/// Helper to convert ENSEMBL and RefSeq gene ID to u32.
pub fn numeric_gene_id(raw_id: &str) -> Result<u32, anyhow::Error> {
    let clean_id = if raw_id.starts_with("ENSG") {
        // Strip "ENSG" prefix and as many zeroes as follow
        raw_id
            .chars()
            .skip("ENSG".len())
            .skip_while(|c| *c == '0')
            .collect()
    } else {
        raw_id.to_owned()
    };

    clean_id
        .parse::<u32>()
        .map_err(|e| anyhow::anyhow!("could not parse gene id {:?}: {}", &clean_id, &e))
}

// The output is wrapped in a Result to allow matching on errors
// Returns an Iterator to the Reader of the lines of the file.
pub fn read_lines<P: AsRef<Path>>(
    filename: P,
) -> std::io::Result<std::io::Lines<std::io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(std::io::BufReader::new(file).lines())
}

/// Select the genome release to use.
#[derive(
    clap::ValueEnum,
    Clone,
    Copy,
    Debug,
    strum::Display,
    PartialEq,
    Eq,
    enum_map::Enum,
    PartialOrd,
    Ord,
    Hash,
)]
pub enum GenomeRelease {
    // GRCh37 / hg19
    #[strum(serialize = "GRCh37")]
    Grch37,
    /// GRCh38 / hg38
    #[strum(serialize = "GRCh38")]
    Grch38,
}

impl GenomeRelease {
    pub fn name(&self) -> String {
        match self {
            GenomeRelease::Grch37 => String::from("GRCh37"),
            GenomeRelease::Grch38 => String::from("GRCh38"),
        }
    }
}

impl From<GenomeRelease> for Assembly {
    fn from(val: GenomeRelease) -> Self {
        match val {
            GenomeRelease::Grch37 => Assembly::Grch37p10,
            GenomeRelease::Grch38 => Assembly::Grch38,
        }
    }
}

impl From<Assembly> for GenomeRelease {
    fn from(assembly: Assembly) -> Self {
        match assembly {
            Assembly::Grch37 | Assembly::Grch37p10 => GenomeRelease::Grch37,
            Assembly::Grch38 => GenomeRelease::Grch38,
        }
    }
}

impl std::str::FromStr for GenomeRelease {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let s = s.to_ascii_lowercase();
        if s.starts_with("grch37") {
            Ok(GenomeRelease::Grch37)
        } else if s.starts_with("grch38") {
            Ok(GenomeRelease::Grch38)
        } else {
            Err(anyhow::anyhow!("Unknown genome release: {}", s))
        }
    }
}

/// The version of `varfish-server-worker` package.
pub const VERSION: &str = env!("CARGO_PKG_VERSION");

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn numeric_gene_id_simple() -> Result<(), anyhow::Error> {
        assert_eq!(1, numeric_gene_id("ENSG0000000001")?);
        assert_eq!(1, numeric_gene_id("ENSG1")?);
        assert_eq!(1, numeric_gene_id("1")?);

        Ok(())
    }
}

#[cfg(test)]
mod test {
    use std::io::Read;

    #[test]
    fn trace_rss_now_smoke() {
        super::trace_rss_now();
    }

    #[test]
    fn build_chrom_map_snapshot() {
        let map = super::build_chrom_map();
        insta::assert_yaml_snapshot!(map);
    }

    #[rstest::rstest]
    #[case(true)]
    #[case(false)]
    fn open_write_maybe_gz(#[case] is_gzip: bool) -> Result<(), anyhow::Error> {
        let filename = if is_gzip { "test.txt" } else { "test.txt.gz" };
        let tmp_dir = temp_testdir::TempDir::default();

        {
            let mut f = super::open_write_maybe_gz(tmp_dir.join(filename))?;
            f.flush()?;
        }

        let mut f = std::fs::File::open(tmp_dir.join(filename)).map(std::io::BufReader::new)?;
        let mut buf = Vec::new();
        f.read_to_end(&mut buf)?;

        insta::assert_snapshot!(format!("{:x?}", &buf));

        Ok(())
    }

    #[rstest::rstest]
    #[case(true)]
    #[case(false)]
    fn open_read_maybe_gz(#[case] is_gzip: bool) -> Result<(), anyhow::Error> {
        let mut f = super::open_read_maybe_gz(if is_gzip {
            "tests/common/test.txt.gz"
        } else {
            "tests/common/test.txt"
        })?;

        let mut buf = String::new();
        f.read_to_string(&mut buf)?;

        insta::assert_snapshot!(&buf);

        Ok(())
    }

    #[rstest::rstest]
    #[case(0..10, 0..10, 1.0)]
    #[case(0..10, 5..15, 0.5)]
    #[case(5..15, 0..10, 0.5)]
    #[case(0..10, 10..20, 0.0)]
    #[case(0..2, 0..10, 0.2)]
    #[case(0..10, 0..2, 0.2)]
    fn reciprocal_overlap(
        #[case] lhs: std::ops::Range<i32>,
        #[case] rhs: std::ops::Range<i32>,
        #[case] expected: f32,
    ) {
        let actual = super::reciprocal_overlap(lhs, rhs);
        assert!(float_cmp::approx_eq!(f32, expected, actual, ulps = 2))
    }

    #[rstest::rstest]
    #[case("ENSG0000000142", 142)]
    #[case("42", 42)]
    fn numeric_gene_id(#[case] raw_id: &str, #[case] expected: u32) {
        let actual = super::numeric_gene_id(raw_id).unwrap();
        assert_eq!(expected, actual);
    }

    #[test]
    fn read_lines() -> Result<(), anyhow::Error> {
        let lines = super::read_lines("tests/common/lines.txt")?.collect::<Result<Vec<_>, _>>()?;

        insta::assert_yaml_snapshot!(lines);

        Ok(())
    }

    #[rstest::rstest]
    #[case(crate::common::GenomeRelease::Grch37, "GRCh37")]
    #[case(crate::common::GenomeRelease::Grch38, "GRCh38")]
    fn genome_release_name(#[case] release: super::GenomeRelease, #[case] expected: &str) {
        assert_eq!(expected, release.name());
    }

    #[rstest::rstest]
    #[case(
        crate::common::GenomeRelease::Grch37,
        hgvs::static_data::Assembly::Grch37p10
    )]
    #[case(
        crate::common::GenomeRelease::Grch38,
        hgvs::static_data::Assembly::Grch38
    )]
    fn assembly_from_genome_release(
        #[case] release: super::GenomeRelease,
        #[case] assembly: hgvs::static_data::Assembly,
    ) -> Result<(), anyhow::Error> {
        let res: hgvs::static_data::Assembly = release.into();

        assert_eq!(res, assembly);

        Ok(())
    }

    #[rstest::rstest]
    #[case(
        crate::common::GenomeRelease::Grch37,
        hgvs::static_data::Assembly::Grch37
    )]
    #[case(
        crate::common::GenomeRelease::Grch37,
        hgvs::static_data::Assembly::Grch37p10
    )]
    #[case(
        crate::common::GenomeRelease::Grch38,
        hgvs::static_data::Assembly::Grch38
    )]
    fn genome_release_from_assembly(
        #[case] release: super::GenomeRelease,
        #[case] assembly: hgvs::static_data::Assembly,
    ) -> Result<(), anyhow::Error> {
        let res: super::GenomeRelease = assembly.into();

        assert_eq!(res, release);

        Ok(())
    }

    #[rstest::rstest]
    #[case(crate::common::GenomeRelease::Grch37, "grch37")]
    #[case(crate::common::GenomeRelease::Grch38, "grch38")]
    fn genome_relese_from_str(
        #[case] release: super::GenomeRelease,
        #[case] s: &str,
    ) -> Result<(), anyhow::Error> {
        let res: super::GenomeRelease = s.parse()?;

        assert_eq!(res, release);

        Ok(())
    }
}
