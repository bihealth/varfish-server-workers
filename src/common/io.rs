//! Common, IO-related code.

use std::{
    fs::File,
    io::{BufRead, BufReader, BufWriter, Write},
    path::Path,
};

use flate2::{bufread::MultiGzDecoder, write::GzEncoder, Compression};

/// Transparently open a file with gzip decoder.
pub fn open_read_maybe_gz<P>(path: P) -> Result<Box<dyn BufRead>, anyhow::Error>
where
    P: AsRef<Path>,
{
    if path.as_ref().extension().map(|s| s.to_str()) == Some(Some("gz")) {
        tracing::trace!("Opening {:?} as gzip for reading", path.as_ref());
        let file = File::open(path)?;
        let bufreader = BufReader::new(file);
        let decoder = MultiGzDecoder::new(bufreader);
        Ok(Box::new(BufReader::new(decoder)))
    } else {
        tracing::trace!("Opening {:?} as plain text for reading", path.as_ref());
        let file = File::open(path).map(BufReader::new)?;
        Ok(Box::new(BufReader::new(file)))
    }
}

/// Transparently open afile with gzip encoder.
pub fn open_write_maybe_gz<P>(path: P) -> Result<Box<dyn Write>, anyhow::Error>
where
    P: AsRef<Path>,
{
    if path.as_ref().extension().map(|s| s.to_str()) == Some(Some("gz")) {
        tracing::trace!("Opening {:?} as gzip for writing", path.as_ref());
        let file = File::create(path)?;
        let bufwriter = BufWriter::new(file);
        let encoder = GzEncoder::new(bufwriter, Compression::default());
        Ok(Box::new(encoder))
    } else {
        tracing::trace!("Opening {:?} as plain text for writing", path.as_ref());
        let file = File::create(path)?;
        Ok(Box::new(file))
    }
}

// The output is wrapped in a Result to allow matching on errors
// Returns an Iterator to the Reader of the lines of the file.
pub fn read_lines<P: AsRef<Path>>(
    filename: P,
) -> std::io::Result<std::io::Lines<std::io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(std::io::BufReader::new(file).lines())
}

#[cfg(test)]
mod test {
    use std::io::Read;

    #[rstest::rstest]
    #[case(true)]
    #[case(false)]
    fn open_write_maybe_gz(#[case] is_gzip: bool) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{:?}", is_gzip);

        let filename = if is_gzip { "test.txt" } else { "test.txt.gz" };
        let tmp_dir = temp_testdir::TempDir::default();

        {
            let mut f = super::open_write_maybe_gz(tmp_dir.join(filename))?;
            f.flush()?;
        }

        let mut f = std::fs::File::open(tmp_dir.join(filename)).map(std::io::BufReader::new)?;
        let mut buf = Vec::new();
        f.read_to_end(&mut buf)?;

        insta::assert_snapshot!(format!("{:x?}", &buf));

        Ok(())
    }

    #[test]
    fn read_lines() -> Result<(), anyhow::Error> {
        let lines = super::read_lines("tests/common/lines.txt")?.collect::<Result<Vec<_>, _>>()?;

        insta::assert_yaml_snapshot!(lines);

        Ok(())
    }
}
