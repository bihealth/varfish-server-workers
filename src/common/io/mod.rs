//! Common, IO-related code.

pub mod std;
