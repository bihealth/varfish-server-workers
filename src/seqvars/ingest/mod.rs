//! Implementation of `seqvars ingest` subcommand.

use crate::common::{self, GenomeRelease};
use noodles_vcf as vcf;

pub mod header;

/// Command line arguments for `seqvars ingest` subcommand.
#[derive(Debug, clap::Parser)]
#[command(author, version, about = "ingest sequence variant VCF", long_about = None)]
pub struct Args {
    /// The assumed genome build.
    #[clap(long)]
    pub genomebuild: GenomeRelease,
    /// Path to input file.
    #[clap(long)]
    pub path_in: String,
    /// Path to output file.
    #[clap(long)]
    pub path_out: String,
}

/// Return the version of the `varfish-server-worker` crate and `x.y.z` in tests.
fn worker_version() -> &'static str {
    if cfg!(test) {
        "x.y.z"
    } else {
        env!("CARGO_PKG_VERSION")
    }
}

/// Main entry point for `seqvars ingest` sub command.
pub fn run(args_common: &crate::common::Args, args: &Args) -> Result<(), anyhow::Error> {
    let before_anything = std::time::Instant::now();
    tracing::info!("args_common = {:#?}", &args_common);
    tracing::info!("args = {:#?}", &args);

    common::trace_rss_now();

    tracing::info!("opening input file...");
    let mut input_reader = {
        let file = std::fs::File::open(&args.path_in)
            .map_err(|e| anyhow::anyhow!("could not open input file {}: {}", &args.path_in, e))
            .map(std::io::BufReader::new)?;
        vcf::reader::Builder::default()
            .build_from_reader(file)
            .map_err(|e| anyhow::anyhow!("could not build VCF reader: {}", e))?
    };

    tracing::info!("analyzing header...");
    let input_header = input_reader
        .read_header()
        .map_err(|e| anyhow::anyhow!("problem reading VCF header: {}", e))?;
    let output_header =
        header::build_output_header(&input_header, args.genomebuild, worker_version())?;

    let mut output_writer = {
        let writer = std::fs::File::create(&args.path_out).map_err(|e| {
            anyhow::anyhow!(
                "could not output file for writing {}: {}",
                &args.path_out,
                e
            )
        })?;
        let writer = std::io::BufWriter::new(writer);
        vcf::writer::Writer::new(writer)
    };
    output_writer.write_header(&output_header)?;

    tracing::info!(
        "All of `seqvars ingest` completed in {:?}",
        before_anything.elapsed()
    );
    Ok(())
}

#[cfg(test)]
mod test {
    use rstest::rstest;

    use crate::common::GenomeRelease;

    macro_rules! set_snapshot_suffix {
        ($($expr:expr),*) => {
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_suffix(format!($($expr,)*));
            let _guard = settings.bind_to_scope();
        }
    }

    #[rstest]
    #[case("tests/seqvars/ingest/example_dragen.07.021.624.3.10.4.vcf")]
    #[case("tests/seqvars/ingest/example_dragen.07.021.624.3.10.9.vcf")]
    #[case("tests/seqvars/ingest/example_gatk_hc.3.7-0.vcf")]
    #[case("tests/seqvars/ingest/example_gatk_hc.4.4.0.0.vcf")]
    fn smoke_test_run(#[case] path: &str) {
        set_snapshot_suffix!("{:?}", path.split('/').last().unwrap().replace(".", "_"));

        let tmpdir = temp_testdir::TempDir::default();

        let args_common = Default::default();
        let args = super::Args {
            genomebuild: GenomeRelease::Grch37,
            path_in: path.into(),
            path_out: tmpdir
                .join("out.vcf")
                .to_str()
                .expect("invalid path")
                .into(),
        };
        super::run(&args_common, &args).unwrap();
    }
}
