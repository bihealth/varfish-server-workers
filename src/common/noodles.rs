//! Helper code for using noodles with S3.
//!
//! At the moment, we only support reading but not yet writing.
//!
//! The main reason is that the awslabs SDK for Rust currently supports easily
//! getting an `AsyncRead` from an S3 object but not an `AsyncWrite`.  For
//! this, we would have to create a wrapper that writes to a multipart upload
//! or similar.

use async_compression::tokio::bufread::GzipDecoder;
use mehari::common::{
    io::{std::is_gz, tokio::open_read_maybe_gz},
    noodles::AsyncVcfReader,
};
use noodles_bgzf as bgzf;
use noodles_core::Position;
use noodles_csi::{self as csi, index::reference_sequence::bin::Chunk};
use noodles_tabix as tabix;
use noodles_vcf as vcf;
use std::{path::Path, pin::Pin};
use tokio::io::{AsyncBufRead, BufReader};

/// Build TBI for file at `path_src` and write to `path_dst`.
pub async fn build_tbi<S, D>(path_src: S, path_dst: D) -> Result<(), anyhow::Error>
where
    S: AsRef<std::path::Path>,
    D: AsRef<std::path::Path>,
{
    let mut reader = tokio::fs::File::open(path_src.as_ref())
        .await
        .map(bgzf::AsyncReader::new)
        .map(vcf::AsyncReader::new)
        .map_err(|e| anyhow::anyhow!("error input file for tbi creation: {}", e))?;

    let header = reader
        .read_header()
        .await
        .map_err(|e| anyhow::anyhow!("error reading header: {}", e))?;

    let mut record = vcf::Record::default();

    let mut indexer = tabix::index::Indexer::default();
    indexer.set_header(csi::index::header::Builder::vcf().build());

    let mut start_position = reader.get_ref().virtual_position();

    while reader
        .read_record(&header, &mut record)
        .await
        .map_err(|e| anyhow::anyhow!("problem reading record: {}", e))?
        != 0
    {
        let end_position = reader.get_ref().virtual_position();
        let chunk = Chunk::new(start_position, end_position);

        let reference_sequence_name = record.chromosome().to_string();
        let start = Position::try_from(usize::from(record.position()))
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))
            .map_err(|e| anyhow::anyhow!("error converting start position: {}", e))?;
        let end = record
            .end()
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))
            .and_then(|position| {
                Position::try_from(usize::from(position))
                    .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))
            })
            .map_err(|e| anyhow::anyhow!("error converting end position: {}", e))?;

        indexer
            .add_record(&reference_sequence_name, start, end, chunk)
            .map_err(|e| anyhow::anyhow!("error adding record to tabix index: {}", e))?;

        start_position = end_position;
    }

    let index = indexer.build();

    {
        let mut writer = tokio::fs::File::create(path_dst.as_ref())
            .await
            .map(tokio::io::BufWriter::new)
            .map(tabix::AsyncWriter::new)
            .map_err(|e| anyhow::anyhow!("error output file for tbi creation: {}", e))?;
        writer
            .write_index(&index)
            .await
            .map_err(|e| anyhow::anyhow!("error writing tabix index: {}", e))?;
        writer
            .shutdown()
            .await
            .map_err(|e| anyhow::anyhow!("error flushing tabix index: {}", e))?;

        tokio::time::sleep(std::time::Duration::from_millis(200)).await;
    }

    Ok(())
}

/// Open plain text or gzip reader via S3.
pub async fn s3_open_read_maybe_gz<P>(path: P) -> Result<Pin<Box<dyn AsyncBufRead>>, anyhow::Error>
where
    P: AsRef<Path>,
{
    let path_string = format!("{}", path.as_ref().display());
    let (bucket, key) = if let Some((bucket, key)) = path_string.split_once('/') {
        (bucket.to_string(), key.to_string())
    } else {
        anyhow::bail!("invalid S3 path: {}", path.as_ref().display());
    };

    let config = aws_config::load_from_env().await;
    let client = aws_sdk_s3::Client::new(&config);
    let object = client.get_object().bucket(&bucket).key(&key).send().await?;

    let path_is_gzip = is_gz(path.as_ref());
    tracing::trace!(
        "Opening S3 object {} as {} for reading (async)",
        path.as_ref().display(),
        if path_is_gzip {
            "gzip (allow multi-member)"
        } else {
            "plain text"
        }
    );

    if path_is_gzip {
        let bufreader = BufReader::new(object.body.into_async_read());
        let decoder = {
            let mut decoder = GzipDecoder::new(bufreader);
            decoder.multiple_members(true);
            decoder
        };
        Ok(Box::pin(BufReader::new(decoder)))
    } else {
        Ok(Box::pin(BufReader::new(object.body.into_async_read())))
    }
}

/// Helper function that opens a list of paths as VCF readers.
pub async fn open_vcf_readers(paths: &[String]) -> Result<Vec<AsyncVcfReader>, anyhow::Error> {
    let mut result = Vec::new();
    for path in paths.iter() {
        let buf_read = if super::s3::s3_mode() && !path.starts_with('/') {
            s3_open_read_maybe_gz(path).await?
        } else {
            open_read_maybe_gz(path).await?
        };
        result.push(AsyncVcfReader::new(buf_read));
    }
    Ok(result)
}

/// Helper function that opens one VCF reader at the given path.
///
/// The behaviour is as follows:
///
/// - If `path_in` is "-" then open stdin and read as plain text.
/// - If environment variable `AWS_PROFILE` is set to "varfish-s3" then enable S3 mode.
/// - If `path_in` is absolute or S3 mode is disabled then open `path_in` as local file
/// - Otherwise, attempt to open `path_in` as S3 object.
pub async fn open_vcf_reader(path_in: &str) -> Result<AsyncVcfReader, anyhow::Error> {
    let s3_mode = match std::env::var("AWS_PROFILE") {
        Ok(s) => s == "varfish-s3",
        _ => false,
    };
    if s3_mode && !path_in.starts_with('/') {
        Ok(vcf::AsyncReader::new(
            s3_open_read_maybe_gz(path_in)
                .await
                .map_err(|e| anyhow::anyhow!("could not build VCF reader from S3 file: {}", e))?,
        ))
    } else {
        Ok(vcf::AsyncReader::new(
            open_read_maybe_gz(path_in).await.map_err(|e| {
                anyhow::anyhow!("could not build VCF reader from local file: {}", e)
            })?,
        ))
    }
}

#[cfg(test)]
mod test {
    #[tokio::test]
    async fn build_tbi() -> Result<(), anyhow::Error> {
        let tmpdir = temp_testdir::TempDir::default();

        let path_in: String = "tests/seqvars/ingest/NA12878_dragen.vcf.gz".into();
        let path_out: String = tmpdir
            .join("out.vcf.gz.tbi")
            .to_str()
            .expect("invalid path")
            .into();
        super::build_tbi(&path_in, &path_out).await?;

        // This must be fixed first https://github.com/zaeleus/noodles/issues/213
        // let mut buffer: Vec<u8> = Vec::new();
        // hxdmp::hexdump(&crate::common::read_to_bytes(&path_out)?, &mut buffer)?;
        // insta::assert_snapshot!(String::from_utf8_lossy(&buffer));

        Ok(())
    }
}
