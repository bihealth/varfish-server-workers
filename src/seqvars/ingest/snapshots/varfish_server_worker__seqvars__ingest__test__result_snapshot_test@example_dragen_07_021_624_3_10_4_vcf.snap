---
source: src/seqvars/ingest/mod.rs
expression: "std::fs::read_to_string(&args.path_out)?"
---
##fileformat=VCFv4.4
##INFO=<ID=gnomad_exomes_an,Number=1,Type=Integer,Description="Number of alleles in gnomAD exomes">
##INFO=<ID=gnomad_exomes_hom,Number=1,Type=Integer,Description="Number of hom. alt. carriers in gnomAD exomes">
##INFO=<ID=gnomad_exomes_het,Number=1,Type=Integer,Description="Number of het. alt. carriers in gnomAD exomes">
##INFO=<ID=gnomad_exomes_hemi,Number=1,Type=Integer,Description="Number of hemi. alt. carriers in gnomAD exomes">
##INFO=<ID=gnomad_genomes_an,Number=1,Type=Integer,Description="Number of alleles in gnomAD genomes">
##INFO=<ID=gnomad_genomes_hom,Number=1,Type=Integer,Description="Number of hom. alt. carriers in gnomAD genomes">
##INFO=<ID=gnomad_genomes_het,Number=1,Type=Integer,Description="Number of het. alt. carriers in gnomAD genomes">
##INFO=<ID=gnomad_genomes_hemi,Number=1,Type=Integer,Description="Number of hemi. alt. carriers in gnomAD genomes">
##INFO=<ID=helix_an,Number=1,Type=Integer,Description="Number of alleles in HelixMtDb">
##INFO=<ID=helix_hom,Number=1,Type=Integer,Description="Number of hom. alt. carriers in HelixMtDb">
##INFO=<ID=helix_het,Number=1,Type=Integer,Description="Number of het. alt. carriers in HelixMtDb">
##INFO=<ID=ANN,Number=.,Type=String,Description="Functional annotations: 'Allele | Annotation | Annotation_Impact | Gene_Name | Gene_ID | Feature_Type | Feature_ID | Transcript_BioType | Rank | HGVS.c | HGVS.p | cDNA.pos / cDNA.length | CDS.pos / CDS.length | AA.pos / AA.length | Distance | ERRORS / WARNINGS / INFO'">
##FILTER=<ID=PASS,Description="All filters passed">
##FORMAT=<ID=AD,Number=R,Type=Integer,Description="Read depth for each allele">
##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read depth">
##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Conditional genotype quality">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=PS,Number=1,Type=Integer,Description="Phase set">
##contig=<ID=1,length=249250621,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=2,length=243199373,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=3,length=198022430,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=4,length=191154276,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=5,length=180915260,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=6,length=171115067,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=7,length=159138663,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=8,length=146364022,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=9,length=141213431,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=10,length=135534747,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=11,length=135006516,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=12,length=133851895,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=13,length=115169878,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=14,length=107349540,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=15,length=102531392,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=16,length=90354753,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=17,length=81195210,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=18,length=78077248,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=19,length=59128983,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=20,length=63025520,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=21,length=48129895,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=22,length=51304566,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=X,length=155270560,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=Y,length=59373566,assembly="GRCh37",species="Homo sapiens">
##contig=<ID=MT,length=16569,assembly="GRCh37",species="Homo sapiens">
##fileDate=20230421
##SAMPLE=<ID=NA12878,Sex="Male",Disease="Affected">
##PEDIGREE=<ID=NA12878>
##x-varfish-case-uuid=00000000-0000-0000-0000-000000000000
##x-varfish-version=<ID=varfish-server-worker,Version="x.y.z">
##x-varfish-version=<ID=orig-caller,Name="Dragen",Version="SW: 07.021.624.3.10.4, HW: 07.021.624">
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	NA12878
17	41256074	.	CA	C	.	.	gnomad_exomes_an=20150;gnomad_exomes_hom=2725;gnomad_exomes_het=5476;gnomad_genomes_an=0;gnomad_genomes_hom=0;gnomad_genomes_het=0;ANN=C|intron_variant|MODIFIER|BRCA1|HGNC:1100|transcript|NM_007294.4|Coding&ManeSelect|6/22|c.441+64delT|p.?|554/7088|441/5592||-64|,C|intron_variant|MODIFIER|BRCA1|HGNC:1100|transcript|NM_007297.4|Coding|5/21|c.300+64delT|p.?|494/7028|300/5451||-64|,C|intron_variant|MODIFIER|BRCA1|HGNC:1100|transcript|NM_007298.3|Coding|5/21|c.441+64delT|p.?|460/3682|441/2280||-64|,C|intron_variant|MODIFIER|BRCA1|HGNC:1100|transcript|NM_007299.4|Coding|6/21|c.441+64delT|p.?|548/3696|441/2100||-64|,C|intron_variant|MODIFIER|BRCA1|HGNC:1100|transcript|NM_007300.4|Coding|6/23|c.441+64delT|p.?|554/7151|441/5655||-64|	GT:AD:DP:GQ:PS	0|1:11,6:17:43:41256074
MT	750	.	A	G	.	.	clinvar_vcv=VCV000441148.2;clinvar_germline_classification=association not found	GT:GQ:AD:DP	1/1:98:1,7818:7819
